### PR TITLE
Add FID to train and val loop

### PR DESCRIPTION
**Pull request by jeswan**

### CODE REVIEW

File: src/metrics/fid.py [hunk 1]
```diff
@@ -68,8 +68,11 @@ def calculate_activation_statistics(batch, model, batch_size=16, dims=2048, devi
     sigma = np.cov(act, rowvar=False)
     return mu, sigma
 
-def calculate_fid(preds, truths, batch_size=16, dims=2048, device='cpu'):
+def calculate_fid(preds, truths, is_bw=True, batch_size=16, dims=2048, device='cpu'):
     """Calculates the FID of two paths"""
+    if is_bw:
+        preds = preds.repeat(1, 3, 1, 1)
+        truths = truths.repeat(1, 3, 1, 1)
```
Comment (jeswan) [inline src/metrics/fid.py:75]: naive way of adding 3 channels (just repeat the intensities across RGB)

File: src/models/model.py [hunk 2]
```diff
@@ -4,15 +4,15 @@
 import torch
 import torch.nn.functional as F
 
-from src.models.unet import UNet
-from src.data.data import DaVinciDataModule
+from models.unet import UNet
+from metrics.fid import calculate_fid
+from data.data import DaVinciDataModule
```
Comment (jeswan) [inline src/models/model.py:9]: Action required: `export PYTHONPATH="${PYTHONPATH}:/home/js11133/capstone/da_vinci_depth/src"` to ~/.bash_profile

File: src/models/model.py [hunk 3]
```diff
@@ -25,6 +25,7 @@ def __init__(
             bilinear: bool = False,
             lr: float = 0.001,
             output_img_freq : int = 100,
+            fid_freq : int = 100,
```
Comment (jeswan) [inline src/models/model.py:29]: This may need to be changed

File: src/models/model.py [hunk 4]
```diff
@@ -72,10 +78,16 @@ def training_step(self, batch, batch_idx):
         self.log('train_loss', loss_val)
         if batch_idx % self.output_img_freq == 0:
             self._log_images(img, target, pred, extra_info, step_name='train')
+        if batch_idx % self.fid_freq == 0:
+            self._log_fid(pred, target)
+
+        start = torch.cuda.Event(enable_timing=True)
+        end = torch.cuda.Event(enable_timing=True)
```
Comment (jeswan) [inline src/models/model.py]:
```suggestion
```

File: src/models/model.py [hunk 5]
```diff
@@ -217,6 +233,6 @@ def add_model_specific_args(parent_parser):
     print("model instance created")
 
     # train
-    trainer = pl.Trainer().from_argparse_args(args)
+    trainer = pl.Trainer.from_argparse_args(args)
```
Comment (jeswan) [inline src/models/model.py:241]: This format is required should we choose to implement callbacks

File: src/models/model.py [hunk 6]
```diff
@@ -62,6 +64,10 @@ def _calc_input_channels(self):
         if self.include_right_view:
             self.input_channels = 2 * self.input_channels
 
+    def _log_fid(self, pred, target):
+        fid_val = calculate_fid(pred, target, batch_size=16, device=self.device)
```
Comment (annikabrundyn) [inline src/models/model.py]: @jeswan I dont think the batch size should be hard coded here - shouldnt it just take self.batch_size?

Comment (jeswan) [reply src/models/model.py]: this batch_size is a little different than the unet batch size. It's the batch size for the Inception network running under the hood of FID (ran into memory errors when it was larger than 16)

Comment (annikabrundyn) [reply src/models/model.py]: does it still work when our batch size is not exactly 16? i.e. 32 or even like 8?

Comment (jeswan) [reply src/models/model.py]: removed since batch_size was not used.

File: src/models/model.py [hunk 7]
```diff
@@ -89,6 +99,9 @@ def validation_step(self, batch, batch_idx):
         # log predicted images
         if batch_idx % self.output_img_freq == 0:
             self._log_images(img, target, pred, extra_info, step_name='valid')
+        #log FID
```
Comment (jeswan) [inline src/models/model.py]:
```suggestion
        # log FID
```

**Pull request merged**